### PR TITLE
feat(examples/templates/gcp-devcontainer): add envbuilder provider

### PR DESCRIPTION
**Pull request by johnstcn**
This PR modifies the gcp-devcontainer example template to include support for devcontainer caching using the envbuilder provider.

### CODE REVIEW

File: examples/templates/gcp-devcontainer/main.tf [hunk 1]
```diff
@@ -144,45 +277,50 @@ resource "google_compute_instance" "vm" {
...
   }
 }
 
-locals {
-  # Ensure Coder username is a valid Linux username
-  linux_user = lower(substr(data.coder_workspace_owner.me.name, 0, 32))
+# Install code-server via Terraform module.
+module "code-server" {
+  count    = data.coder_workspace.me.start_count
+  source   = "https://registry.coder.com/modules/code-server"
```
Comment (matifali) [inline examples/templates/gcp-devcontainer/main.tf:319]:
```suggestion
  source   = "registry.coder.com/modules/code-server/coder"
  version  = "1.0.17"
```

**Review by matifali — APPROVED** [review]
A minor suggestion to use the latest module version. Otherwise looks good to me. 
Thank you for doing it 😊

File: examples/templates/gcp-devcontainer/main.tf [hunk 2]
```diff
@@ -6,16 +6,49 @@ terraform {
...
+}
+
+variable "insecure_cache_repo" {
+  default     = false
+  description = "Enable this option if your cache registry does not serve HTTPS."
+  type        = bool
+}
+
+variable "cache_repo_docker_config_path" {
+  default     = ""
```
Comment (mtojek) [inline examples/templates/gcp-devcontainer/main.tf:46]: Is there any standard path we can suggest here or mention in the description?

File: examples/templates/gcp-devcontainer/main.tf [hunk 3]
```diff
@@ -51,25 +85,48 @@ data "coder_parameter" "zone" {
...
-  name  = "coder-${data.coder_workspace.me.id}-root"
-  type  = "pd-ssd"
-  image = "debian-cloud/debian-12"
-  lifecycle {
-    ignore_changes = [name, image]
-  }
+data "coder_parameter" "devcontainer_builder" {
+  description  = <<-EOF
+Image that will build the devcontainer.
+We highly recommend using a specific release as the `:latest` tag will change.
```
Comment (mtojek) [inline examples/templates/gcp-devcontainer/main.tf:122]: Maybe rephrase it to indicate it is dangerous? For example: Do not use `envbuilder:latest` tag due to the risk of build instability.

File: examples/templates/gcp-devcontainer/main.tf [hunk 4]
```diff
@@ -80,46 +137,122 @@ data "coder_parameter" "repo_url" {
...
-    display_name = "Disk Usage"
-    interval     = 5
-    timeout      = 5
-    script       = "coder stat disk"
+  # If we have a cached image, use the cached image's environment variables. Otherwise, just use
+  # the environment variables we've defined above.
+  docker_env_input = try(envbuilder_cached_image.cached.0.env_map, local.envbuilder_env)
+  # Convert the above to the list of arguments for the Docker run command. This is going to end
+  # up in our startup script metadata. These are all terminated by backslashes.
+  docker_env_arg_list = [for k, v in local.docker_env_input : " -e \"${k}=${v}\" \\"]
```
Comment (mtojek) [inline examples/templates/gcp-devcontainer/main.tf:180]: Should user modify these locals? If not, maybe indicate that in a comment.

File: examples/templates/gcp-devcontainer/main.tf [hunk 5]
```diff
@@ -6,16 +6,49 @@ terraform {
...
   description = "Which Google Compute Project should your workspace live in?"
 }
 
+variable "cache_repo" {
+  default     = ""
+  description = "(Optional) Use a container registry as a cache to speed up builds."
+  type        = string
+}
+
+variable "insecure_cache_repo" {
```
Comment (mtojek) [inline examples/templates/gcp-devcontainer/main.tf:39]: This changeset is relatively big, so is the `.tf` file. Maybe wipe out this variable?

File: examples/templates/gcp-devcontainer/main.tf [hunk 6]
```diff
@@ -6,16 +6,49 @@ terraform {
...
+}
+data "coder_workspace_owner" "me" {}
+
 variable "project_id" {
   description = "Which Google Compute Project should your workspace live in?"
 }
 
+variable "cache_repo" {
+  default     = ""
+  description = "(Optional) Use a container registry as a cache to speed up builds."
```
Comment (mtojek) [inline examples/templates/gcp-devcontainer/main.tf:35]: People may be struggling whether `cache_repo` is a path, URI, endpoint, etc.

File: examples/templates/gcp-devcontainer/main.tf [hunk 7]
```diff
@@ -6,16 +6,49 @@ terraform {
     google = {
       source = "hashicorp/google"
     }
+    envbuilder = {
```
Comment (mtojek) [inline examples/templates/gcp-devcontainer/main.tf:9]: curious: will we support the original form without `envbuilder`? if so, should we add a new example like `gcp-devcontainer`?

File: examples/templates/gcp-devcontainer/main.tf [hunk 8]
```diff
@@ -6,16 +6,49 @@ terraform {
...
 
+provider "google" {
+  zone    = data.coder_parameter.zone.value
+  project = var.project_id
+}
+
+data "google_compute_default_service_account" "default" {}
+
+data "coder_workspace" "me" {
+}
```
Comment (mafredri) [inline examples/templates/gcp-devcontainer/main.tf:26]:
```suggestion
data "coder_workspace" "me" {}
```
Consistency.

File: examples/templates/gcp-devcontainer/main.tf [hunk 9]
```diff
@@ -51,25 +85,48 @@ data "coder_parameter" "zone" {
...
-  type  = "pd-ssd"
-  image = "debian-cloud/debian-12"
-  lifecycle {
-    ignore_changes = [name, image]
-  }
+data "coder_parameter" "devcontainer_builder" {
+  description  = <<-EOF
+Image that will build the devcontainer.
+We highly recommend using a specific release as the `:latest` tag will change.
+Find the latest version of Envbuilder here: https://github.com/coder/envbuilder/pkgs/container/envbuilder
```
Comment (mafredri) [inline examples/templates/gcp-devcontainer/main.tf:123]:
```suggestion
Find the latest version of Envbuilder here: https://ghcr.io/coder/envbuilder
```
Shorter URL, maybe nicer?

File: examples/templates/gcp-devcontainer/main.tf (hunk 4 above)
Comment (mafredri) [inline examples/templates/gcp-devcontainer/main.tf:180]: This conversion might have trouble with some inputs, writing to an env file and referencing that is an alternative as it doesn't require quoting. Newline inputs could still spell trouble though.

File: examples/templates/gcp-devcontainer/main.tf [hunk 10]
```diff
@@ -80,46 +137,122 @@ data "coder_parameter" "repo_url" {
...
+     --net=host \
+     -h ${lower(data.coder_workspace.me.name)} \
+     -v /home/${local.linux_user}/envbuilder:/workspaces \
+     -v /var/run/docker.sock:/var/run/docker.sock \
+    ${join("\n", local.docker_env_arg_list)}
+      ${data.coder_parameter.devcontainer_builder.value}
+  META
+}
+
+# Create a persistent to store the workspace data.
```
Comment (mafredri) [inline examples/templates/gcp-devcontainer/main.tf:225]:
```suggestion
# Create a persistent disk to store the workspace data.
```

File: examples/templates/gcp-devcontainer/main.tf [hunk 11]
```diff
@@ -144,45 +277,50 @@ resource "google_compute_instance" "vm" {
...
-      echo "${local.linux_user} ALL=(ALL) NOPASSWD:ALL" > /etc/sudoers.d/coder-user
-    fi
+# Create a Coder agent to manage the workspace.
+resource "coder_agent" "dev" {
+  count              = data.coder_workspace.me.start_count
+  arch               = "amd64"
+  auth               = "token"
+  os                 = "linux"
+  dir                = "/workspaces/${trimsuffix(basename(data.coder_parameter.repo_url.value), ".git")}"
+  connection_timeout = 0
```
Comment (mafredri) [inline examples/templates/gcp-devcontainer/main.tf:291]: Does setting this to zero actually serve a purpose? Doesn't it just use the default of 30s in this case?

**Review by mafredri — APPROVED** [review]
Nice work 🎉

File: examples/templates/gcp-devcontainer/main.tf (hunk 9 above)
Comment (johnstcn) [reply examples/templates/gcp-devcontainer/main.tf:123]: Oh nice, TIL you could do this!

File: examples/templates/gcp-devcontainer/main.tf (hunk 11 above)
Comment (johnstcn) [reply examples/templates/gcp-devcontainer/main.tf:291]: Unsure; this was from the original template IIRC.

File: examples/templates/gcp-devcontainer/main.tf (hunk 7 above)
Comment (johnstcn) [reply examples/templates/gcp-devcontainer/main.tf:9]: The `envbuilder` provider only gets used if `cache_repo` is set.

File: examples/templates/gcp-devcontainer/main.tf (hunk 4 above)
Comment (johnstcn) [reply examples/templates/gcp-devcontainer/main.tf:179]: Something like this?

`echo <base64-encoded> | base64 -d > env.list`

File: examples/templates/gcp-devcontainer/main.tf (hunk 2 above)
Comment (johnstcn) [reply examples/templates/gcp-devcontainer/main.tf:45]: I'll mention `~/.docker/config.json` but it'll depend heavily on their setup.

File: examples/templates/gcp-devcontainer/main.tf (hunk 4 above)
Comment (johnstcn) [reply examples/templates/gcp-devcontainer/main.tf:179]: yeah probably not :D
